src/avram/attribute.cr

```diff
@@ -1,77 +1,69 @@
 class Avram::Attribute(T)
   alias ErrorMessage = String | Proc(String, String, String) | Avram::CallableErrorMessage
-  getter :name
-  setter :value
-  getter :param_key
+  getter name : Symbol
+  setter value : T?
+  getter param_key : String
   @errors = [] of String
+  @param : Avram::Uploadable | String?
 
-  def initialize(
-    @name : Symbol,
-    @param : Avram::Uploadable | String?,
-    @value : T,
-    @param_key : String
-  )
+  def initialize(@name, @param, @value : T?, @param_key)
     @original_value = @value
   end
 
   @_permitted : Avram::PermittedAttribute(T)?
 
   def permitted
     @_permitted ||= begin
-      Avram::PermittedAttribute.new(name: @name, param: @param, value: @value, param_key: @param_key).tap do |attribute|
+      Avram::PermittedAttribute(T).new(name: @name, param: @param, value: @value, param_key: @param_key).tap do |attribute|
         errors.each do |error|
           attribute.add_error error
         end
       end
     end
   end
 
-  def param
+  def param : Avram::Uploadable | String
     @param || value.to_s
   end
 
-  def add_error(message : ErrorMessage = "is invalid")
-    perform_add_error(message)
-  end
-
-  private def perform_add_error(message : String = "is invalid")
+  def add_error(message : String = "is invalid")
     @errors << message
   end
 
-  private def perform_add_error(message : (Proc | Avram::CallableErrorMessage))
```

Comment:
Moving this overload into the `add_error` method simplifies the code and error stack traces